### DISCUSSION TITLE
how to build windows app from macos?

### DISCUSSION THREAD

**Discussion by zgid123** [tauri-apps/tauri#9650] [Q&A]
Hi,

I am using macOS 14 and tauri v1.

I already followed this post https://tauri.app/v1/guides/building/cross-platform to install all necessary things to build windows. But when running the script to run build, it raises error

```
error: failed to run custom build command for `my-app v0.0.1 (/Users/alpha/Alpha/personal/products/my-app/workspaces/tauri-app/src-tauri)`

Caused by:
  process didn't exit successfully: `/Users/alpha/Alpha/personal/products/my-app/target/release/build/my-app-6f0f7adb963a857b/build-script-build` (exit status: 1)
  --- stdout
  cargo:rerun-if-env-changed=TAURI_CONFIG
  cargo:rerun-if-changed=tauri.conf.json
  cargo:rustc-cfg=desktop
  package.metadata does not exist
  cargo:warning=Compiler family detection failed due to error: ToolExecError: Command "clang-cl" "-E" "/Users/alpha/Alpha/personal/products/my-app/target/x86_64-pc-windows-msvc/release/build/my-app-d82842a77614be08/out/5086315667597211200detect_compiler_family.c" with args clang-cl did not execute successfully (status code exit status: 1).
  cargo:warning=clang-cl: error: unsupported option '-x c'; did you mean '/TC' or '/TP'?

  --- stderr


  error occurred: Command "clang-cl" "-nologo" "-MD" "-O2" "-Brepro" "-m64" "-I" "/Users/alpha/Alpha/personal/products/my-app/target/x86_64-pc-windows-msvc/release/build/my-app-d82842a77614be08/out" "--target=x86_64-pc-windows-msvc" "-Wno-unused-command-line-argument" "-fuse-ld=lld-link" "/imsvc/Users/alpha/Library/Caches/cargo-xwin/xwin/crt/include" "/imsvc/Users/alpha/Library/Caches/cargo-xwin/xwin/sdk/include/ucrt" "/imsvc/Users/alpha/Library/Caches/cargo-xwin/xwin/sdk/include/um" "/imsvc/Users/alpha/Library/Caches/cargo-xwin/xwin/sdk/include/shared" "-xc" "-DRC_INVOKED" "-E" "--" "/Users/alpha/Alpha/personal/products/my-app/target/x86_64-pc-windows-msvc/release/build/my-app-d82842a77614be08/out/resource.rc" with args clang-cl did not execute successfully (status code exit status: 1).
```

I am not sure how to fix this one.

**Comment by FabianLars** [comment]
Hi, can you share the output of `clang --version` (or was it `clang-cl --version` ? 🤔 )? If it's higher than 15, try to downgrade if possible.

**Reply by zgid123** [reply]
my version is 15.0.0

```
Apple clang version 15.0.0 (clang-1500.1.0.2.5)
Target: arm64-apple-darwin23.4.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

lemme try to downgrade it

**Reply by FabianLars** [reply]
hmm, no need. I use the very same thing on my machine without issues. Can you create a new empty project via create-tauri-app and check if that has the same issue?

**Reply by zgid123** [reply]
it raises the same issue

**Reply by perry-ruan** [reply]
i have the same issue
`Apple clang version 15.0.0 (clang-1500.3.9.4)
Target: arm64-apple-darwin23.3.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin`

**Reply by WriteNaN** [reply]
same here, how did it go OP?

**Reply by gabenddos** [reply]
UP
Same issue here

**Reply by simonhyll** [reply]
There's already a ⬆️ reaction you can use, just commenting you're experiencing the issue as well isn't all that useful, especially as a response to someone 's message because now we're poking Fabian needlessly as responses to his comment instead of as new comments on the actual thread 😅

**Reply by gabenddos** [reply]
Got it, I'm willing to help, just don't know how to.

I'm not really sure if this is related but, I found this issue on llvm project but I'm clueless if the problem is in Tauri or on a dependency. 
https://github.com/llvm/llvm-project/issues/63930

Is this a bug or do we have an issue related to this error? I mean, if it's not external, cause in this case I think there isn't much we could do.

Gonna post the full output of the command with the --verbose flag
```
...
✓ built in 4.94s
    Running [tauri_cli] Command `cargo-xwin  build --bins --features tauri/custom-protocol,tauri/native-tls --release --target x86_64-pc-windows-msvc`
   Compiling app v0.1.0 (/Users/gabrielrodrigues/dev/ar-worship/src-tauri)
The following warnings were emitted during compilation:

warning: app@0.1.0: Compiler family detection failed due to error: ToolExecError: Command "clang-cl" "-E" "/Users/gabrielrodrigues/dev/ar-worship/src-tauri/target/x86_64-pc-windows-msvc/release/build/app-c15a41c2fed294d7/out/4589244665001407914detect_compiler_family.c" with args clang-cl did not execute successfully (status code exit status: 1).
warning: app@0.1.0: clang-cl: error: unsupported option '-x c'; did you mean '/TC' or '/TP'?

error: failed to run custom build command for `app v0.1.0 (/Users/gabrielrodrigues/dev/ar-worship/src-tauri)`

Caused by:
  process didn't exit successfully: `/Users/gabrielrodrigues/dev/ar-worship/src-tauri/target/release/build/app-a321befb9303fb92/build-script-build` (exit status: 1)
  --- stdout
  cargo:rerun-if-env-changed=TAURI_CONFIG
  cargo:rerun-if-changed=tauri.conf.json
  cargo:rustc-cfg=desktop
  cargo:rustc-env=TAURI_ANDROID_PACKAGE_PREFIX=com_gabrilucido
  cargo:PERMISSION_FILES_PATH=/Users/gabrielrodrigues/dev/ar-worship/src-tauri/target/x86_64-pc-windows-msvc/release/build/app-c15a41c2fed294d7/out/app-manifest/__app__-permission-files
  cargo:rerun-if-changed=capabilities
  cargo:rustc-env=TAURI_ENV_TARGET_TRIPLE=x86_64-pc-windows-msvc
  package.metadata does not exist
  cargo:warning=Compiler family detection failed due to error: ToolExecError: Command "clang-cl" "-E" "/Users/gabrielrodrigues/dev/ar-worship/src-tauri/target/x86_64-pc-windows-msvc/release/build/app-c15a41c2fed294d7/out/4589244665001407914detect_compiler_family.c" with args clang-cl did not execute successfully (status code exit status: 1).
  cargo:warning=clang-cl: error: unsupported option '-x c'; did you mean '/TC' or '/TP'?

  --- stderr


  error occurred: Command "clang-cl" "-nologo" "-MD" "-O2" "-Brepro" "-m64" "-I" "/Users/gabrielrodrigues/dev/ar-worship/src-tauri/target/x86_64-pc-windows-msvc/release/build/app-c15a41c2fed294d7/out" "--target=x86_64-pc-windows-msvc" "-Wno-unused-command-line-argument" "-fuse-ld=lld-link" "/imsvc/Users/gabrielrodrigues/Library/Caches/cargo-xwin/xwin/crt/include" "/imsvc/Users/gabrielrodrigues/Library/Caches/cargo-xwin/xwin/sdk/include/ucrt" "/imsvc/Users/gabrielrodrigues/Library/Caches/cargo-xwin/xwin/sdk/include/um" "/imsvc/Users/gabrielrodrigues/Library/Caches/cargo-xwin/xwin/sdk/include/shared" "-xc" "-DRC_INVOKED" "-E" "--" "/Users/gabrielrodrigues/dev/ar-worship/src-tauri/target/x86_64-pc-windows-msvc/release/build/app-c15a41c2fed294d7/out/resource.rc" with args clang-cl did not execute successfully (status code exit status: 1).


    Error [tauri_cli_node] failed to build app: failed to build app
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Reply by FabianLars** [reply]
> especially as a response to someone 's message because now we're poking Fabian needlessly as responses to his comment instead of as new comments on the actual thread 😅

I get notified either way 🤣 

But yeah, no idea what to do here and since it's still working fine on my machines (including VMs with different mac versions) i can't even mindlessly play around with it :(

One thing to try is to remove _everything_ that may be reladed to clang or llvm that was installed via homebrew so that there's really only the xcode clang installation left.

Also would be cool if you could check if compiling a non-tauri rust app still works with the cargo-xwin instructions (from our docs or theirs)

**Reply by cliqer** [reply]
You must be missing the `LDFLAGS` and `CPPFLAGS`.
If using OS's clang, `echo $LDFLAGS` should return an empty string; if not, it is set in `~/.zshrc`.
Otherwise, try installing a newer llvm or llvm@17 with brew.
```cmd
brew install llvm

echo 'export PATH="/opt/homebrew/opt/llvm/bin:$PATH"' >> ~/.zshrc
echo 'export LDFLAGS="-L/opt/homebrew/opt/llvm/lib"' >> ~/.zshrc
echo 'export CPPFLAGS="-I/opt/homebrew/opt/llvm/include"' >> ~/.zshrc

source ~/.zshrc
```
If llvm@18 errors, downgrade to llvm@17 or lower
```cmd
brew uninstall llvm
brew install llvm@17

echo 'export PATH="/opt/homebrew/opt/llvm@17/bin:$PATH"' >> ~/.zshrc
export LDFLAGS="-L/opt/homebrew/opt/llvm@17/lib"
export CPPFLAGS="-I/opt/homebrew/opt/llvm@17/include"
```
`cargo clean --manifest-path src-tauri/Cargo.toml`

for `x64 Windows`:
```cmd
rustup target add x86_64-pc-windows-msvc
pnpm tauri build --runner cargo-xwin --target x86_64-pc-windows-msvc
```
for `aarch64 Windows`:
```cmd
rustup target add aarch64-pc-windows-msvc
pnpm tauri build --runner cargo-xwin --target aarch64-pc-windows-msvc
```
In `/src-tauri/build.rs`
```rs
#[cfg(target_os = "windows")]
fn main() {
    let mut windows = tauri_build::WindowsAttributes::new();

    windows = windows.app_manifest(
        r#"
        <assembly xmlns="urn:schemas-microsoft-com:asm.v1" manifestVersion="1.0">
        <dependency>
            <dependentAssembly>
            <assemblyIdentity
                type="win32"
                name="Microsoft.Windows.Common-Controls"
                version="6.0.0.0"
                processorArchitecture="*"
                publicKeyToken="6595b64144ccf1df"
                language="*"
            />
            </dependentAssembly>
        </dependency>
        <trustInfo xmlns="urn:schemas-microsoft-com:asm.v3">
            <asmv3:windowsSettings xmlns="http://schemas.microsoft.com/SMI/2024/WindowsSettings">
                <supportedArchitectures>amd64 arm64</supportedArchitectures>
            </asmv3:windowsSettings>
            <security>
                <requestedPrivileges xmlns="urn:schemas-microsoft-com:asm.v3">
                    <requestedExecutionLevel level="requireAdministrator" uiAccess="false" />
                </requestedPrivileges>
            </security>
        </trustInfo>
        </assembly>
        "#,
    );

    tauri_build::try_build(tauri_build::Attributes::new().windows_attributes(windows))
        .expect("failed to run build script");
}

#[cfg(not(target_os = "windows"))]
fn main() {
    tauri_build::build();
}
```

**Reply by ragoso** [reply]
UP
Same issue here

**Reply by kkhai** [reply]
> > especially as a response to someone 's message because now we're poking Fabian needlessly as responses to his comment instead of as new comments on the actual thread 😅
> 
> I get notified either way 🤣
> 
> But yeah, no idea what to do here and since it's still working fine on my machines (including VMs with different mac versions) i can't even mindlessly play around with it :(
> 
> One thing to try is to remove _everything_ that may be reladed to clang or llvm that was installed via homebrew so that there's really only the xcode clang installation left.
> 
> Also would be cool if you could check if compiling a non-tauri rust app still works with the cargo-xwin instructions (from our docs or theirs)

I removed the llvm package from brew and it works. Thank you.

**Comment by websmurf** [comment]
I ran into one after the other problem when trying to build a Tauri app on Mac OS, ended up creating a docker container which enabled me to get a successful build.

If you still have issues, you could try this image: https://hub.docker.com/repository/docker/websmurf/tauri-builder/general

**Comment by akiyaani2** [comment]
## Amanda Torres — Discussion #736 Actionable Items → GitHub Issues

Per PM mandate, converting actionable items from this discussion to tracked GitHub Issues.

**Issues Created:**
- **#799** — Per-Agent Budget Enforcement at Cron Layer (11/11 divisions flagged HIGH PRIORITY)
- **#800** — GCP MCP Server — Enable for Data + Engineering Agents  
- **#801** — Task Checkout Atomicity for Concurrent Agents (Helena flagged CRITICAL for trading)
- **#802** — Skills Audit: Install Missing Tier-1 Skills Across All Divisions

*Weekly review of this discussion is now on my PM cadence.*

— Amanda Torres, CoS | Program Management
